src/instruct.jl

```diff
@@ -64,25 +29,28 @@ function _prepare_instruct(state, U, locs::NTuple{M}, control_locs, control_bits
 end
 
 function YaoBase.instruct!(
-    state::AbstractVecOrMat{T},
-    operator::AbstractMatrix{T},
+    state::AbstractVecOrMat{T1},
+    operator::AbstractMatrix{T2},
     locs::Tuple{},
-    control_locs::NTuple{C, Int} = (),
-    control_bits::NTuple{C, Int} = ()) where {T, M, C}
+    control_locs::NTuple{C, Int}=(),
+    control_bits::NTuple{C, Int}=()) where {T1, T2, M, C}
     return state
 end
 
 function YaoBase.instruct!(
-    state::AbstractVecOrMat{T},
-    operator::AbstractMatrix{T},
+    state::AbstractVecOrMat{T1},
+    operator::AbstractMatrix{T2},
     locs::NTuple{M, Int},
-    control_locs::NTuple{C, Int} = (),
-    control_bits::NTuple{C, Int} = ()) where {T, M, C}
-
-    U = sort_unitary(operator, locs)
-    locs_raw, ic = _prepare_instruct(state, U, locs, control_locs, control_bits)
+    control_locs::NTuple{C, Int}=(),
+    control_bits::NTuple{C, Int}=()) where {T1, T2, M, C}
 
-    return _instruct!(state, autostatic(U), locs_raw, ic)
+    if T2!=T1
+        @warn "Element Type Mismatch: register $(T1), operator $(T2). Converting operator to match, this may cause performance issue"
+        operator = copyto!(similar(operator, T1), operator)
+    end
+    operator = sort_unitary(operator, locs)
+    locs_raw, ic = _prepare_instruct(state, operator, locs, control_locs, control_bits)
+    return _instruct!(state, autostatic(operator), locs_raw, ic)
```

Comment:
I think you need to transpose the operator and state for general case in order to make memory contiguous for BLAS?

Reply:
state is transposed, operator is not nessesary. After all, we don't have BLAS...